metadetect/metadetect_and_cal.py

```diff
@@ -0,0 +1,208 @@
+import copy
+import ngmix
+import numpy as np
+import esutil as eu
+import galsim
+
+from . import detect
+from . import fitting
+
+STEP = 0.01
+SHEARS = {
+    '1p': galsim.Shear(g1=STEP, g2=0.0),
+    '1m': galsim.Shear(g1=-STEP, g2=0.0),
+    '2p': galsim.Shear(g1=0.0, g2=STEP),
+    '2m': galsim.Shear(g1=0.0, g2=-STEP)}
+
+
+def do_metadetect_and_cal(
+        config, mbobs, rng, wcs_func=None, psf_rec_funcs=None):
+    """
+    Meta-detect and cal on the multi-band observations.
+    """
+    md = MetadetectAndCal(
+        config, mbobs, rng, wcs_func=wcs_func, psf_rec_funcs=psf_rec_funcs)
+    md.go()
+    return md.result
+
+
+class MetadetectAndCal(dict):
+    """
+    Meta-detect and cal on the multi-band observations.
+
+    parameters
+    ----------
+    config: dict
+        Configuration dictionary. Possible entries are
+            metacal
+            weight (if calculating weighted moments)
+            max: (if running a max like fitter)
+            fofs (if running MOF)
+            mof (if running MOF)
+
+    """
+    def __init__(self, config, mbobs, rng, wcs_func=None, psf_rec_funcs=None):
+        self._set_config(config)
+        self.mbobs = mbobs
+        self.nband = len(mbobs)
+        self.rng = rng
+        self.wcs_func = wcs_func
+        if psf_rec_funcs is None:
+            self.psf_rec_funcs = [None] * len(mbobs)
+        else:
+            assert len(psf_rec_funcs) == len(mbobs)
+            self.psf_rec_funcs = psf_rec_funcs
+
+        self._set_fitter()
+
+    @property
+    def result(self):
+        """
+        get the result dict, keyed by the metacal type such
+        as 'noshear', '1p', '1m', '2p', '2m'
+        """
+        if not hasattr(self, '_result'):
+            raise RuntimeError('run go() first')
+
+        return self._result
+
+    def go(self):
+        """
+        make sheared versions of the images, run detection and measurements
+        on each
+        """
+        odict = self._get_all_metacal()
+
+        self._result = {}
+        for key, sheared_mbobs in odict.items():
+            self._result[key] = self._measure(sheared_mbobs, key)
+
+    def _measure(self, sheared_mbobs, mcal_step):
+        """
+        perform measurements on the input mbobs. This involves running
+        detection as well as measurements
+        """
+
+        if mcal_step == 'noshear':
+            pos_transform = None
+        else:
+            ainv = np.linalg.inv(SHEARS[mcal_step].getMatrix())
+            dims = sheared_mbobs[0][0].image.shape
+            x_cen = (dims[0] - 1) / 2
+            y_cen = (dims[1] - 1) / 2
+
+            def pos_transform(x, y):
+                x = np.atleast_1d(x) - x_cen
+                y = np.atleast_1d(y) - y_cen
+                out = np.dot(ainv, np.vstack([x, y]))
+                assert out.shape[1] == x.shape[0]
+                return out[0, :] + x_cen, out[1, :] + y_cen
```

Comment:
@esheldon Here is what I have for unshearing the positions. Basically, according to Mike and galsim docs, the shear is applied around the "true center" of the interpolated image. This routine takes sextractor positions, subtracts the center, undoes the shear, and adds the center back. 

The only thing missing might be the position offset if the `sep` python package still uses 1-based indexing. However, looking at the code already in the repo you seemed to assume it was reporting zero-based indexes.